### PR TITLE
[docs] Add log correlation docs

### PR DESCRIPTION
**Pull request by gregkalapos**
Documenting our log correlation story for .NET - lots of ideas are stolen from the great [Go doc](https://github.com/elastic/apm-agent-go/blob/master/docs/log-correlation.asciidoc).

Context: 
- We plan to release 2 new NuGet packages, `Elastic.Apm.NLog` and `Elastic.Apm.SerilogEnricher`, they are not yet released, but I already linked them in this docs. 
- Those 2 packages live in the [`ecs-dotnet`](https://github.com/elastic/ecs-dotnet/) repo with the rest of the ecs .NET things.
- I'm trying to document here how users can setup those packages, plus also cover manual trace injections.

### CODE REVIEW

File: docs/log-correlation.asciidoc [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+* Manually inject trace IDs into log records, as described below in <<log-correlation-manual>>.
+
+Regardless of how you integrate APM with logging, you can use {filebeat-ref}[Filebeat] to
+send your logs to Elasticsearch, in order to correlate your traces and logs and link from
+APM to the {apm-app-ref}/xpack-logs.html[Logs app].
+
+[[serilog]]
+== Serilog
+
+We offer a https://github.com/serilog/serilog/wiki/Enrichment[Serilog Enricher] that adds the trace id to every log line that is created during an active trace.
```
Comment (gregkalapos) [inline docs/log-correlation.asciidoc]: I'm not sure if this external link is needed. I just don't want to assume that everyone knows what a serilog enricher is - on the other hand, serilog users probably will know it. 

Maybe it's useful if someone is just researching logging frameworks and wants to know what we support. 

But if the external link is a no-go, I'm ok to remove. 

Same applies to `LayoutRenderers` for NLog below.

File: docs/log-correlation.asciidoc [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+[source,json]
+----
+{
+  "description": "...",
+  "processors": [
+    {
+      "grok": {
+        "field": "message",
+        "patterns": [%{LOGLEVEL:log.level} \\[trace.id=%{TRACE_ID:trace.id}(?: transaction.id=%{SPAN_ID:transaction.id})?\\] %{GREEDYDATA:message}"],
```
Comment (gregkalapos) [inline docs/log-correlation.asciidoc]: This needs to be reviewed - I removed some things compared to the go docs. I have ~very limited~ none grok knowledge.

File: docs/log-correlation.asciidoc [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+* Open a feature request, or contribute code, for additional support, as described in https://github.com/elastic/apm-agent-dotnet/blob/master/CONTRIBUTING.md[CONTRIBUTING.md].
+* Manually inject trace IDs into log records, as described below in <<log-correlation-manual>>.
+
+Regardless of how you integrate APM with logging, you can use {filebeat-ref}[Filebeat] to
+send your logs to Elasticsearch, in order to correlate your traces and logs and link from
+APM to the {apm-app-ref}/xpack-logs.html[Logs app].
+
+[[serilog]]
+== Serilog
```
Comment (bmorelli25) [inline docs/log-correlation.asciidoc]: This renders as a top level heading right now. Screen Shot 2019-12-18 at 11 18 37 AM

Let's nest instead.
```suggestion
=== Serilog
```

File: docs/log-correlation.asciidoc [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+  CustomFormatter = new EcsTextFormatter()
+})
+.CreateLogger();
+``` 
+
+With this setup the application will send all the logs automatically to Elasticsearch and you will be able to jump from traces to logs and from logs to traces.
+
+
+[[nlog]]
+== NLog
```
Comment (bmorelli25) [inline docs/log-correlation.asciidoc]:
```suggestion
=== NLog
```

File: docs/log-correlation.asciidoc [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+endif::[]
+
+[[log-correlation]]
+== Log correlation
+
+
+The Elastic APM .NET agent provides integrations for popular logging frameworks, which take care of
+injecting trace ID fields into your application's log records. Currently supported logging frameworks are:
+
+- Serilog
```
Comment (bmorelli25) [inline docs/log-correlation.asciidoc]:
```suggestion
- <<serilog>>
```

File: docs/log-correlation.asciidoc [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+- NLog
```
Comment (bmorelli25) [inline docs/log-correlation.asciidoc]:
```suggestion
- <<nlog>>
```

File: docs/log-correlation.asciidoc [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+
+If your favorite logging framework is not already supported, there are two other options:
+
+* Open a feature request, or contribute code, for additional support, as described in https://github.com/elastic/apm-agent-dotnet/blob/master/CONTRIBUTING.md[CONTRIBUTING.md].
+* Manually inject trace IDs into log records, as described below in <<log-correlation-manual>>.
```
Comment (bmorelli25) [inline docs/log-correlation.asciidoc]:
```suggestion
* Manually inject trace IDs into log records, as described in <<log-correlation-manual>>.
```

File: docs/log-correlation.asciidoc [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+</targets>
+<rules>
+    <logger name="*" minLevel="Trace" writeTo="logfile" />
+</rules>
+</nlog>
+```
+
+As you can see in the sample file above, you can reference the current transaction id with `${ElasticApmTransactionId}` and the trace id with `${ElasticApmTraceId}`.
+
+[float]
```
Comment (bmorelli25) [inline docs/log-correlation.asciidoc]: As of right now, this is nested under nLog. Let's add this to its own page.
```suggestion
```

File: docs/log-correlation.asciidoc [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+[[log-correlation-manual-structured]]
+==== Manual log correlation (structured)
+
+For correlating structured logs with traces, the following fields should be added to your logs:
+
+ - `trace.id`
+ - `transaction.id`
+ 
+Given a transaction object, you can obtain its trace id by using the `Transaction.TraceId` property and its transaction id by using the `Transaction.Id` property
```
Comment (bmorelli25) [inline docs/log-correlation.asciidoc]:
```suggestion
Given a transaction object, you can obtain its trace id by using the `Transaction.TraceId` property and its transaction id by using the `Transaction.Id` property.
```

File: docs/log-correlation.asciidoc [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+
+
+[[nlog]]
+== NLog
+
+For NLog, we offer two https://github.com/NLog/NLog/wiki/Layout-Renderers[LayoutRenderers] that inject the current trace and transaction id into logs. 
+
+In order to use them, you need to add the https://www.nuget.org/packages/Elastic.Apm.NLog[Elastic.Apm.NLog] NuGet package to your application and load it in the `<extensions>` section of your NLog config file:
+
+```
```
Comment (bmorelli25) [inline docs/log-correlation.asciidoc]: Let's change this from a large code block to a structured code block so it renders in the same format as the other code on the page. The backticks are pretty ugly:
Screen Shot 2019-12-18 at 11 29 56 AM


```suggestion
[source,xml]
----
```

File: docs/log-correlation.asciidoc [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+        <attribute name="traceid" layout="${ElasticApmTraceId}" />
+        <attribute name="transactionid" layout="${ElasticApmTransactionId}" />
+    </layout>
+</target>
+</targets>
+<rules>
+    <logger name="*" minLevel="Trace" writeTo="logfile" />
+</rules>
+</nlog>
+```
```
Comment (bmorelli25) [inline docs/log-correlation.asciidoc]:
```suggestion
----
```

File: docs/log-correlation.asciidoc [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+----
+
+In case the agent is configured and there is an active transaction, the `traceId` and `transactionId` will always return the current trace and transaction ids that you can manually add to your logs. Make sure you store those in the fields `trace.id` and `transaction.id` when you send them to Elasticsearch.
+
+[float]
+[[log-correlation-manual-unstructured]]
+==== Manual log correlation (unstructured)
+
+For correlating unstructured logs (e.g. basic printf-style logging, like
+`Console.WriteLine`), then you will need to include the trace ids in your log message, and then
```
Comment (bmorelli25) [inline docs/log-correlation.asciidoc]:
```suggestion
`Console.WriteLine`), you will need to include the trace ids in your log message, and then
```

File: docs/log-correlation.asciidoc [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+----
+var currentTransaction = //Get Current transaction, e.g.: Agent.Tracer.CurrentTransaction;
+
+Console.WriteLine($"ERROR [trace.id={currentTransaction.TraceId} transaction.id={currentTransaction.Id}] an error occurred");
+----
+
+
+This would print a log message along the lines of:
+
+    ERROR [trace.id=cd04f33b9c0c35ae8abe77e799f126b7 transaction.id=cd04f33b9c0c35ae] an error occurred
```
Comment (bmorelli25) [inline docs/log-correlation.asciidoc]:
```suggestion
----
    ERROR [trace.id=cd04f33b9c0c35ae8abe77e799f126b7 transaction.id=cd04f33b9c0c35ae] an error occurred
----
```

File: docs/log-correlation.asciidoc (hunk 2 above)
Comment (bmorelli25) [reply docs/log-correlation.asciidoc]: Haha, sounds good. I'll give this a test and see if it works

**Comment by gregkalapos** [comment]
> Here's a first round of comments

Thanks, I went through it - all fine, pushed them.

**Comment by bmorelli25** [comment]
I used the unstructured log message you provided:
```
ERROR [trace.id=cd04f33b9c0c35ae8abe77e799f126b7 transaction.id=cd04f33b9c0c35ae] an error 
```
And with the grok pattern you provided, I was able to return this json:
```json
{
  "trace": {
    "id": "cd04f33b9c0c35ae8abe77e799f126b7"
  },
  "log": {
    "level": "ERROR"
  },
  "message": "an error occurred",
  "transaction": {
    "id": "cd04f33b9c0c35ae"
  }
}
```

**Comment by gregkalapos** [comment]
> I used the unstructured log message you provided:
> 
> ```
> ERROR [trace.id=cd04f33b9c0c35ae8abe77e799f126b7 transaction.id=cd04f33b9c0c35ae] an error 
> ```
> 
> And with the grok pattern you provided, I was able to return this json:
> 
> ```json
> {
>   "trace": {
>     "id": "cd04f33b9c0c35ae8abe77e799f126b7"
>   },
>   "log": {
>     "level": "ERROR"
>   },
>   "message": "an error occurred",
>   "transaction": {
>     "id": "cd04f33b9c0c35ae"
>   }
> }
> ```

Thank you! That's exactly what it should be.

**Comment by bmorelli25** [comment]
Awesome! Is the `transaction.id` optional? Because that pattern works with `transaction.id` omitted as well.

**Comment by gregkalapos** [comment]
> Awesome! Is the `transaction.id` optional? Because that pattern works with `transaction.id` omitted as well.

From what I know - it is. In fact, on the UI I think we only use `trace.id` in the current version.

File: docs/log-correlation.asciidoc (hunk 1 above)
Comment (bmorelli25) [reply docs/log-correlation.asciidoc]: I think the links make sense. Let's keep them.

**Pull request merged**